providers/frame/polkadot_dev.ts

```diff
@@ -16,12 +27,84 @@ export class PolkadotDevProvider extends FrameBinProvider {
     })
   }
 
+  override async chainFile(pathInfo: PathInfo): Promise<File> {
+    const url = new URL(fromPathInfo({ ...pathInfo, filePath: "user_i" }), this.env.href)
+      .toString()
+    const file = await super.chainFile(pathInfo)
+    return new File(`
+      ${file.codeRaw}
+
+      type ArrayOfLength<
+        T,
+        L extends number,
+        A extends T[] = [],
+      > = number extends L ? T[]
+        : L extends A["length"] ? A
+        : ArrayOfLength<T, L, [...A, T]>
+
+      export async function users<N extends number>(count: N): Promise<ArrayOfLength<C.Sr25519, N>>
+      export async function users(count: number): Promise<C.Sr25519[]> {
+        const response = await fetch(
+          ${JSON.stringify(url)},
+          {
+            method: "POST",
+            headers: {
+              "content-type": "application/json",
+            },
+            body: JSON.stringify({ count }),
+          },
+        )
+        if (!response.ok) {
+          throw new Error(await response.text())
+        }
+        const { index }: { index: number } = await response.json()
+        if (index === -1) {
+          throw new Error("Maximum test user count reached")
+        }
+        const userIds: C.Sr25519[] = []
+        for (let i = index; i < index + count; i++) {
+          userIds.push(C.testUser(i))
+        }
+        return userIds
+      }
+    `)
+  }
+
+  override async handle(request: Request, pathInfo: PathInfo): Promise<Response> {
+    if (pathInfo.filePath === "user_i") {
+      const { count } = await request.json()
+      if (!(count > 0)) {
+        throw new Error("'count' should be greater than 0")
+      }
+      let index = this.#userCount[pathInfo.target!] ?? 0
+      const newCount = index + count
+      if (newCount < PolkadotDevProvider.DEFAULT_TEST_USER_COUNT) {
+        this.#userCount[pathInfo.target!] = newCount
+      } else {
+        index = -1
+      }
+      return new Response(
+        JSON.stringify({ index }),
+        {
+          status: 200,
+          headers: {
+            "content-type": "application/json",
+          },
+        },
+      )
+    }
+    return super.handle(request, pathInfo)
+  }
+
   async launch(pathInfo: PathInfo) {
-    const runtimeName = pathInfo.target
+    const runtimeName = pathInfo.target || "polkadot"
```

Comment:
If there's no target, it should error.
```suggestion
    const runtimeName = pathInfo.target
```

Reply:
this is optional in the `PathInfo` which is parsed using this regex

If I follow correctly, `target` is optional in the regex too

https://github.com/paritytech/capi/blob/3aee893724fd73ff451986bd6eb2c21fd2b42f24/server/PathInfo.ts#L1-L11

Reply:
nvm, I got the suggestion